### PR TITLE
Internal: Create Nested base infrastructure [ED-9199]

### PR DESCRIPTION
**Pull request by davseve**
## PR Checklist

- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [x ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

File: modules/nested-tabs/assets/js/frontend/handlers/nested-tabs.js [hunk 1]
```diff
@@ -7,28 +7,7 @@ export default class NestedTabs extends BaseNestedTabs {
...
-
-				++index;
-			} );
-
-			// On refresh since indexes are rearranged, do not call `activateDefaultTab` let editor control handle it.
-			if ( $removed.length ) {
-				return elementorModules.ViewModule.prototype.onInit.apply( this, args );
-			}
-		}
+		this.addCollapseClassToItems( args );
```
Comment (davseve) [inline modules/nested-tabs/assets/js/frontend/handlers/nested-tabs.js:10]: This Function is needed in `nested-tabs` and `mega-menu`. 
But in `nested-tabs` it must run in the child class. 
That is why is in the parent class but is called only in the child class.

File: modules/nested-tabs/assets/js/editor/index.js [hunk 2]
```diff
@@ -1,6 +1,5 @@
-elementorCommon.elements.$window.on( 'elementor:init-components', async () => {
+elementorCommon.elements.$window.on( 'elementor/nested-element-type-loaded', async () => {
 	// The module should be loaded only when `nestedElements` is available.
-	await elementor.modules.nestedElements;
```
Comment (udidol) [inline modules/nested-tabs/assets/js/editor/index.js:3]: Fix/remove comment, remove empty line

File: modules/nested-tabs/assets/js/editor/module.js [hunk 3]
```diff
@@ -5,4 +5,3 @@ export default class Module {
 		elementor.elementsManager.registerElementType( new NestedTabs() );
 	}
 }
-
```
Comment (udidol) [inline modules/nested-tabs/assets/js/editor/module.js:8]: ?

File: modules/nested-elements/assets/js/frontend/handlers/base-nested.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
+import Base from 'elementor-frontend/handlers/base';
+
+export default class BaseNested extends Base {
```
Comment (udidol) [inline modules/nested-elements/assets/js/frontend/handlers/base-nested.js:3]: Move contents of this class to BaseNestedTabs and delete the class and its imports and usage

File: assets/dev/js/frontend/modules.js [hunk 5]
```diff
@@ -3,7 +3,7 @@ import Document from './document';
 import StretchElement from './tools/stretch-element';
 import BaseHandler from './handlers/base';
 import SwiperBase from './handlers/base-swiper';
-import BaseNestedTabs from 'elementor-frontend/handlers/base-nested-tabs';
+import BaseNestedTabs from 'elementor/modules/nested-tabs/assets/js/base-nested-tabs';
```
Comment (udidol) [inline assets/dev/js/frontend/modules.js:6]: revert move

File: modules/nested-tabs/assets/js/base-nested-tabs.js [hunk 6]
```diff
@@ -1,6 +1,6 @@
-import Base from './base';
+import BaseNested from 'elementor/modules/nested-elements/assets/js/frontend/handlers/base-nested';
```
Comment (udidol) [inline modules/nested-tabs/assets/js/base-nested-tabs.js:1]: revert

File: modules/nested-tabs/assets/js/base-nested-tabs.js [hunk 7]
```diff
@@ -61,13 +37,16 @@ export default class BaseNestedTabs extends Base {
 		};
 	}
 
-	activateDefaultTab() {
+	activateDefaultTab( openDefaultTab = true ) {
```
Comment (udidol) [inline modules/nested-tabs/assets/js/base-nested-tabs.js:40]: Remove this param, in mega menu check for edit mode before calling it.

File: modules/nested-tabs/assets/js/base-nested-tabs.js [hunk 8]
```diff
@@ -176,37 +155,28 @@ export default class BaseNestedTabs extends Base {
...
-				event.preventDefault();
-				this.changeActiveTab( event.currentTarget.getAttribute( 'data-tab' ), true );
-			},
-		} );
+			let index = 1;
+
+			this.findElement( '.e-con' ).each( function() {
+				const $current = jQuery( this ),
+					$desktopTabTitle = $widget.find( `.e-n-tabs-heading > *:nth-child(${ index })` ),
+					mobileTitleHTML = `<div class="e-n-tab-title e-collapse" data-tab="${ index }" role="tab">${ $desktopTabTitle.html() }</div>`;
```
Comment (udidol) [inline modules/nested-tabs/assets/js/base-nested-tabs.js:168]: Generalize class names, pass as variables in an `args` object.
E.g.:
```
args = {
	headingClassName: 'e-n-tabs-heading',
	titleClassName: 'e-n-tab-title'
};
```

File: modules/nested-tabs/assets/js/frontend/handlers/nested-tabs.js [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import BaseNestedTabs from 'elementor-frontend/handlers/base-nested-tabs';
+import BaseNestedTabs from '../../base-nested-tabs';
```
Comment (udidol) [inline modules/nested-tabs/assets/js/frontend/handlers/nested-tabs.js:1]: Revert

**Review by udidol — CHANGES_REQUESTED** [review]

File: modules/nested-tabs/assets/js/editor/index.js (hunk 2 above)
Comment (davseve) [reply modules/nested-tabs/assets/js/editor/index.js:3]: Done

File: modules/nested-tabs/assets/js/editor/module.js (hunk 3 above)
Comment (davseve) [reply modules/nested-tabs/assets/js/editor/module.js:8]: Done

File: modules/nested-elements/assets/js/frontend/handlers/base-nested.js (hunk 4 above)
Comment (davseve) [reply modules/nested-elements/assets/js/frontend/handlers/base-nested.js:3]: Done

File: assets/dev/js/frontend/modules.js (hunk 5 above)
Comment (davseve) [reply assets/dev/js/frontend/modules.js:6]: Done

File: modules/nested-tabs/assets/js/base-nested-tabs.js (hunk 6 above)
Comment (davseve) [reply modules/nested-tabs/assets/js/base-nested-tabs.js:1]: Done

File: modules/nested-tabs/assets/js/base-nested-tabs.js (hunk 7 above)
Comment (davseve) [reply modules/nested-tabs/assets/js/base-nested-tabs.js:40]: I have a problem here, the `activateDefaultTab` runs in the parent class `onInit()` therefore bounding the function in a condition won't do any difference. 
I'd like to hear your suggestion regarding this one.

File: modules/nested-tabs/assets/js/frontend/handlers/nested-tabs.js (hunk 9 above)
Comment (davseve) [reply modules/nested-tabs/assets/js/frontend/handlers/nested-tabs.js:1]: Done

File: modules/nested-tabs/assets/js/editor/module.js (hunk 3 above)
Comment (davseve) [reply modules/nested-tabs/assets/js/editor/module.js:8]: I don't understand why we still see this change.

File: modules/nested-tabs/assets/js/base-nested-tabs.js (hunk 7 above)
Comment (davseve) [reply modules/nested-tabs/assets/js/base-nested-tabs.js]: Done

File: modules/nested-tabs/assets/js/base-nested-tabs.js (hunk 8 above)
Comment (davseve) [reply modules/nested-tabs/assets/js/base-nested-tabs.js]: Took the data from the `setting` object

File: assets/dev/scss/helpers/variables.scss [hunk 10]
```diff
@@ -19,12 +19,14 @@ $gray-bg:       #e6e9ec;
...
 $blue-flat:     #1b607f;
 $blue-light:    #5bc0de;
+$cyan:          #58D0F5;
 $green:         #5cb85c;
 $green-light:   #2ecc71;
 $orange:        #f0ad4e;
 $red:           #d9534f;
 $purple:		#9C7AC1;
 
+
```
Comment (udidol) [inline assets/dev/scss/helpers/variables.scss:29]:
```suggestion
```

File: assets/dev/js/frontend/modules.js [hunk 11]
```diff
@@ -3,7 +3,7 @@ import Document from './document';
 import StretchElement from './tools/stretch-element';
 import BaseHandler from './handlers/base';
 import SwiperBase from './handlers/base-swiper';
-import BaseNestedTabs from 'elementor-frontend/handlers/base-nested-tabs';
+import NestedTabs from '../../../../modules/nested-tabs/assets/js/frontend/handlers/nested-tabs';
```
Comment (udidol) [inline assets/dev/js/frontend/modules.js:6]:
```suggestion
import NestedTabs from 'elementor/modules/nested-tabs/assets/js/frontend/handlers/nested-tabs';
```

File: assets/dev/js/frontend/preloaded-modules.js [hunk 12]
```diff
@@ -8,7 +8,7 @@ import Video from './handlers/video';
 import ImageCarousel from './handlers/image-carousel';
 import TextEditor from './handlers/text-editor';
 
-import NestedTabs from 'elementor/modules/nested-tabs/assets/js/frontend/handlers/nested-tabs';
+import NestedTabs from '../../../../modules/nested-tabs/assets/js/frontend/handlers/nested-tabs';
```
Comment (udidol) [inline assets/dev/js/frontend/preloaded-modules.js:11]:
```suggestion
import NestedTabs from 'elementor/modules/nested-tabs/assets/js/frontend/handlers/nested-tabs';
```

File: assets/dev/js/frontend/elements-handlers-manager.js [hunk 13]
```diff
@@ -27,7 +27,7 @@ module.exports = function( $ ) {
 	};
 
 	if ( elementorFrontendConfig.experimentalFeatures[ 'nested-elements' ] ) {
-		this.elementsHandlers[ 'nested-tabs.default' ] = () => import( /* webpackChunkName: 'nested-tabs' */ 'elementor/modules/nested-tabs/assets/js/frontend/handlers/nested-tabs' );
+		this.elementsHandlers[ 'nested-tabs.default' ] = () => import( /* webpackChunkName: 'nested-tabs' */ '../../../../modules/nested-tabs/assets/js/frontend/handlers/nested-tabs.js' );
```
Comment (udidol) [inline assets/dev/js/frontend/elements-handlers-manager.js:30]:
```suggestion
		this.elementsHandlers[ 'nested-tabs.default' ] = () => import( /* webpackChunkName: 'nested-tabs' */ 'elementor/modules/nested-tabs/assets/js/frontend/handlers/nested-tabs' );
```

**Pull request merged**